pennylane/ops/qubit/special_unitary.py

```diff
@@ -0,0 +1,346 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule contains the operation SpecialUnitary and
+its utility functions.
+"""
+# pylint: disable=arguments-differ
+from functools import lru_cache, reduce
+from itertools import product
+
+import numpy as np
+import pennylane as qml
+from pennylane.operation import AnyWires, Operation
+
+_pauli_matrices = np.array(
+    [[[1, 0], [0, 1]], [[0, 1], [1, 0]], [[0, -1j], [1j, 0]], [[1, 0], [0, -1]]]
+)
+"""Single-qubit Paulis:    I                 X                   Y                  Z"""
+
+_pauli_letters = "IXYZ"
+"""Single-qubit Pauli letters that make up Pauli words."""
+
+
+@lru_cache
+def pauli_basis(num_wires):
```

Comment:
Yeah, I was thinking about this as well. As we didn't go for integration of `SpecialUnitary` with `PauliWord` etc yet, I thought I'd keep them separate. But maybe it's indeed better to move them.
There also is some similar, lazy-evaluating generator in `ArbitraryUnitary` ...

Reply:
I moved `pauli_basis` to `pauli/utils.py` , but it feels a bit strange.
There are a lot of functionalities in that file, and most are concerned with converting/generating in formats
- binary vector representations of Paulis
- `qml.PauliWord`/`qml.Tensor`
- strings

whereas these methods are concerned with the dense matrix  and strings representations, respectively.

Reply:
great point, appreciate that reasoning. perhaps the Pauli module needs more organizing/planning before we dump more stuff into it, especially if we want it to become a multi-tool of sorts. I'm still thinking about it, but I'm considering it outside the scope of this PR

Reply:
@dwierichs are these functions general enough that you think they may be useful beyond `SpecialUnitary`? Part of the reason behind `qml.Pauli` was that we noticed Pauli logic was being replicated _everywhere_ throughout the codebase, because many developers didn't realize other (specialized) functions included it.

@timmysilv is there a way of keeping track of external Pauli logic for now?

Reply:
I totally agree that there is a considerable amount of duplications.
I'm just afraid of putting more into that module that differs _slightly_ (e.g. by being lazy in evaluation or not and by the used input/output formats). I think there is indeed some organizational work required, and the two methods added here have the following features that make me think it might be ok to leave them in `special_unitary.py`:
- They are very short
- They have limited scope: They are very fast for few qubits but actually refuse to compute for medium or many qubits.
- They only do minimal input validation.

Reply:
Makes sense!

Reply:
I'll resolve then, alright? @josh146 @timmysilv